### PR TITLE
feat: Jamba 1.5 Support

### PR DESCRIPTION
**Pull request by amirai21**

### CODE REVIEW

File: ai21/version.py [hunk 1]
```diff
@@ -1 +1 @@
-VERSION = "2.12.0"
```
Comment (amirai21) [inline ai21/version.py:1]: @asafgardin - should be fixed prior merge manually or it's auto generated?

File: examples/studio/chat/chat_response_format.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+class ZooTicket(BaseModel):
+    ticket_type: TicketType
+    quantity: int
+
+
+class ZooTicketsOrder(BaseModel):
+    date: str
+    tickets: list[ZooTicket]
+
+
```
Comment (amirai21) [inline examples/studio/chat/chat_response_format.py:26]: Needs adaptation...

File: README.md [hunk 3]
```diff
@@ -270,6 +271,36 @@ completion_response = client.completion.create(
...
+    max_tokens=100,
+    temperature=0.7,
+    top_p=1.0,
+    stop=["\n"],
+)
+
+print(response)
+```
+
+Note that jamba-instruct supports async streaming as well.
```
Comment (Josephasafg) [inline README.md:302]: "async and streaming"

File: ai21/clients/studio/resources/chat/chat_completions.py [hunk 4]
```diff
@@ -17,13 +21,16 @@ class ChatCompletions(StudioResource, BaseChatCompletions):
...
+        messages: List[ChatMessageParam],
         max_tokens: int | NotGiven = NOT_GIVEN,
         temperature: float | NotGiven = NOT_GIVEN,
         top_p: float | NotGiven = NOT_GIVEN,
         stop: str | List[str] | NotGiven = NOT_GIVEN,
         n: int | NotGiven = NOT_GIVEN,
         stream: Optional[Literal[False]] | NotGiven = NOT_GIVEN,
+        tools: List[ToolDefinition] | NOT_GIVEN = NOT_GIVEN,
+        response_format: ResponseFormat | NOT_GIVEN = NOT_GIVEN,
+        documents: List[DocumentSchema] | NOT_GIVEN = NOT_GIVEN,
```
Comment (Josephasafg) [inline ai21/clients/studio/resources/chat/chat_completions.py:33]: Fix typing for `NotGiven`

File: ai21/clients/studio/resources/chat/chat_completions.py [hunk 5]
```diff
@@ -32,27 +39,33 @@ def create(
...
+        messages: List[ChatMessageParam],
         stream: Literal[True],
         max_tokens: int | NotGiven = NOT_GIVEN,
         temperature: float | NotGiven = NOT_GIVEN,
         top_p: float | NotGiven = NOT_GIVEN,
         stop: str | List[str] | NotGiven = NOT_GIVEN,
         n: int | NotGiven = NOT_GIVEN,
+        tools: List[ToolDefinition] | NOT_GIVEN = NOT_GIVEN,
+        response_format: ResponseFormat | NOT_GIVEN = NOT_GIVEN,
+        documents: List[DocumentSchema] | NOT_GIVEN = NOT_GIVEN,
```
Comment (Josephasafg) [inline ai21/clients/studio/resources/chat/chat_completions.py:51]: Fix typing for `NotGiven`

File: ai21/models/chat/__init__.py [hunk 6]
```diff
@@ -14,4 +20,17 @@
...
+    "DocumentSchema",
+    "FunctionToolDefinition",
+    "ResponseFormat",
+    "ToolCall",
+    "ToolDefinition",
+    "ToolFunction",
+    "ToolParameters",
 ]
+
+from .tool_parameters import ToolParameters
```
Comment (Josephasafg) [inline ai21/models/chat/__init__.py:36]: Move import above `__all__`

File: ai21/version.py (hunk 1 above)
Comment (Josephasafg) [reply ai21/version.py:1]: auto

File: examples/studio/chat/chat_function_calling.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+from ai21 import AI21Client
+from ai21.logger import set_verbose
+from ai21.models.chat import ChatMessage, ToolMessage
+from ai21.models.chat.function_tool_definition import FunctionToolDefinition
+from ai21.models.chat.tool_defintions import ToolDefinition
+from ai21.models.chat.tool_parameters import ToolParameters
```
Comment (Josephasafg) [inline examples/studio/chat/chat_function_calling.py:6]: Import from chat and not all the way from the module itself

File: examples/studio/chat/chat_function_calling.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+messages = [
+    ChatMessage(
+        role="system",
+        content="You are a helpful customer support assistant. Use the supplied tools to assist the user.",
+    ),
+    ChatMessage(role="user", content="Hi, can you tell me the delivery date for my order?"),
+    ChatMessage(role="assistant", content="Hi there! I can help with that. Can you please provide your order ID?"),
+    ChatMessage(role="user", content="i think it is order_12345"),
+]
```
Comment (Josephasafg) [inline examples/studio/chat/chat_function_calling.py:24]: Use new chat message objects

File: examples/studio/chat/chat_function_calling_multiple_tools.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+response = client.chat.completions.create(messages=messages, model="jamba-1.5-large", tools=tools)
+
+print(response.choices[0].message)
+
+assistant_message = response.choices[0].message
+messages.append(assistant_message)
+
+tool_calls = assistant_message.tool_calls
+
+too_call_id_to_result = {}
```
Comment (Josephasafg) [inline examples/studio/chat/chat_function_calling_multiple_tools.py:80]: Lets try to make this part of the examples more readable

File: examples/studio/chat/chat_response_format.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+messages = [
+    ChatMessage(
+        role="user",
+        content="Please create a JSON object for ordering zoo tickets for September 22, 2024, "
+        f"for myself and two kids, based on the following JSON schema: {ZooTicketsOrder.model_json_schema()}.",
+    )
+]
+
+client = AI21Client(api_host="https://api-stage.ai21.com", api_key="F6iFeKlMsisusyhtoy1ZUj4bRPhEd6sf")
```
Comment (Josephasafg) [inline examples/studio/chat/chat_response_format.py:35]: REMOVE and REVOKE

File: tests/unittests/clients/studio/resources/chat/test_chat_completions.py [hunk 11]
```diff
@@ -1,14 +1,26 @@
...
+from ai21.models.chat.tool_parameters import ToolParameters
 
 _DUMMY_API_KEY = "dummy_api_key"
 
 
 def test_chat_create__when_bad_import_to_chat_message__raise_error():
     with pytest.raises(ValueError) as e:
         AI21Client(api_key=_DUMMY_API_KEY).chat.completions.create(
-            model="jamba-instruct-preview",
+            model="jamba-1.5",
```
Comment (Josephasafg) [inline tests/unittests/clients/studio/resources/chat/test_chat_completions.py:23]: Rename jamba-1.5 to either mini or large

File: tests/unittests/clients/studio/resources/chat/test_chat_completions.py [hunk 12]
```diff
@@ -35,10 +47,204 @@ async def test_async_chat_create__when_bad_import_to_chat_message__raise_error()
...
+            "index": 0,
+            "message": {"role": "assistant", "content": " Hello! How can I assist you today?"},
+            "finish_reason": "stop",
+        }
+    ],
+    "usage": {"prompt_tokens": 144, "completion_tokens": 14, "total_tokens": 158},
+    "meta": {"requestDurationMillis": 236},
+}
+
+basic_happy_flow_expected_content = " Hello! How can I assist you today?"
```
Comment (Josephasafg) [inline tests/unittests/clients/studio/resources/chat/test_chat_completions.py:76]: We use terminology like - `_DUMMY_HAPPY_FLOW_RESPONSE_JSON` for this kind of testing contsts

File: tests/unittests/clients/studio/resources/chat/test_chat_completions.py [hunk 13]
```diff
@@ -35,10 +47,204 @@ async def test_async_chat_create__when_bad_import_to_chat_message__raise_error()
...
+    client = AI21Client(api_key=_DUMMY_API_KEY, http_client=mocked_client)
+    response: ChatCompletionResponse = client.chat.completions.create(
+        model="jamba-1.5", messages=[UserMessage(role="user", content="Hello")]
+    )
+    assert response.choices[0].message.content == basic_happy_flow_expected_content
+
+
+@pytest.mark.asyncio
+async def test_async_chat_completion_basic_happy_flow(mocker: MockerFixture) -> None:
+    mocked_client = mocker.Mock(spec=httpx.Client)
```
Comment (Josephasafg) [inline tests/unittests/clients/studio/resources/chat/test_chat_completions.py:91]: `httpx.AsyncClient`

File: tests/integration_tests/clients/test_studio.py [hunk 14]
```diff
@@ -48,6 +49,7 @@
         "when_summarize_by_segment__should_return_ok",
         "when_tokenization__should_return_ok",
         "when_chat_completions__should_return_ok",
+        "when_chat_completions_jamba_instruct__should_return_ok",
```
Comment (Josephasafg) [inline tests/integration_tests/clients/test_studio.py:52]: Lets add the examples you added as well

File: examples/studio/chat/chat_function_calling.py [hunk 15]
```diff
@@ -1,15 +1,17 @@
...
+
 from ai21 import AI21Client
 from ai21.logger import set_verbose
 from ai21.models.chat import ChatMessage, ToolMessage
-from ai21.models.chat.function_tool_definition import FunctionToolDefinition
-from ai21.models.chat.tool_defintions import ToolDefinition
-from ai21.models.chat.tool_parameters import ToolParameters
+from ai21.models.chat import FunctionToolDefinition
+from ai21.models.chat import ToolDefinition
+from ai21.models.chat import ToolParameters
```
Comment (Josephasafg) [inline examples/studio/chat/chat_function_calling.py:8]: They can all be imported in one line

File: examples/studio/chat/chat_function_calling_multiple_tools.py [hunk 16]
```diff
@@ -66,49 +71,57 @@ def get_sunset_hour(place: str, date: str):
...
+
         else:
             print(f"Unexpected tool call found - {tool_call.function.name}")
 else:
-    print("No tool calls found.")
+    print("No tool calls found")
 
 
 if too_call_id_to_result:
+    """Continue the conversation by passing the sunset and weather back to the AI model:"""
```
Comment (Josephasafg) [inline examples/studio/chat/chat_function_calling_multiple_tools.py:120]: Instead of AI model, lets rename to Jamba X

**Review by Josephasafg — APPROVED** [review]

**Pull request merged**